### ISSUE TITLE
R140 - Parse JSON string from AJAX.LastData returns null

### ISSUE THREAD

**Issue by el3um4s** [Scirra/Construct-3-bugs#2689]
## Problem description

Parse JSON string from AJAX.LastData doesn't work.

## Attach a .c3p

[New project.c3p.zip](https://github.com/Scirra/Construct-3-bugs/files/2931625/New.project.c3p.zip)

## Steps to reproduce

Open the attached file and run the preview

## Observed result

![cattura](https://user-images.githubusercontent.com/23579630/53816913-c1126b00-3f64-11e9-9e23-db54b8ae88da.PNG)

## Expected result

![cattura-2](https://user-images.githubusercontent.com/23579630/53816951-d8e9ef00-3f64-11e9-948d-1f9de0f24a1d.PNG)

## More details

**Affected browsers/platforms:** Chrome / Desktop (Windows)

**First affected release:** first started happening in r140

## System details

<details><summary>Chrome</summary>

Platform information
Browser: Chrome
Browser version: 72.0.3626.121
Browser engine: Blink
Browser architecture: 64-bit
Context: browser
Operating system: Windows
Operating system version: 10
Operating system architecture: 64-bit
Device type: desktop
Device pixel ratio: 1
Logical CPU cores: 4
Approx. device memory: 4 GB
User agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/72.0.3626.121 Safari/537.36
C3 release: r140 (beta)
Language setting: en-US

Local storage
Storage quota (approx): 15 gb
Storage usage (approx): 472 mb (3.1%)
Persistant storage: No

Browser support notes
This list contains missing features that are not required, but could improve performance or user experience if supported.

Nothing is missing. Everything is OK!
WebGL information
Version string: WebGL 2.0 (OpenGL ES 3.0 Chromium)
Numeric version: 2
Supports NPOT textures: yes
Supports GPU profiling: yes
Supports highp precision: yes
Vendor: Google Inc.
Renderer: ANGLE (Intel(R) HD Graphics 4000 Direct3D11 vs_5_0 ps_5_0)
Major performance caveat: no
Maximum texture size: 16384
Point size range: 1 to 1024
Extensions:

EXT_color_buffer_float
EXT_disjoint_timer_query_webgl2
EXT_texture_filter_anisotropic
OES_texture_float_linear
WEBGL_compressed_texture_s3tc
WEBGL_compressed_texture_s3tc_srgb
WEBGL_debug_renderer_info
WEBGL_debug_shaders
WEBGL_lose_context
Audio information
System sample rate: 44100 Hz
Output channels: 2
Output interpretation: speakers
Supported decode formats:

Ogg Opus (audio/ogg; codecs=opus)
WebM Opus (audio/webm; codecs=opus)
Ogg Vorbis (audio/ogg; codecs=vorbis)
WebM Vorbis (audio/webm; codecs=vorbis)
MPEG-4 AAC (audio/mp4; codecs=mp4a.40.5)
MP3 (audio/mpeg)
FLAC (audio/flac)
PCM WAV (audio/wav; codecs=1)
Supported encode formats:

WebM Opus (audio/webm; codecs=opus)
Video information
Supported decode formats:

WebM VP9 (video/webm; codecs=vp9)
WebM VP8 (video/webm; codecs=vp8)
Ogg Theora (video/ogg; codecs=theora)
H.264 (video/mp4; codecs=avc1.42E01E)
Supported encode formats:

WebM VP9 (video/webm; codecs=vp9)
WebM VP8 (video/webm; codecs=vp8)

</details>

<details><summary>Windows</summary>

Platform information
Browser: NW.js
Browser version: 0.35.1
Browser engine: Chromium
Browser architecture: 64-bit
Context: nwjs
Operating system: Windows
Operating system version: 10
Operating system architecture: 64-bit
Device type: desktop
Device pixel ratio: 1
Logical CPU cores: 4
Approx. device memory: 4 GB
User agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (@ac9418ba9c3bd7f6baaffa0b055dfe147e0f8364) NWjs/0.35.1 (like Chrome/71.0.3578.80)
C3 release: r140 (beta)
Language setting: en-US

Local storage
Storage quota (approx): 173 gb
Storage usage (approx): 103 mb (0.1%)
Persistant storage: Yes

Browser support notes
This list contains missing features that are not required, but could improve performance or user experience if supported.

Nothing is missing. Everything is OK!
WebGL information
Version string: WebGL 2.0 (OpenGL ES 3.0 Chromium)
Numeric version: 2
Supports NPOT textures: yes
Supports GPU profiling: yes
Supports highp precision: yes
Vendor: Google Inc.
Renderer: ANGLE (Intel(R) HD Graphics 4000 Direct3D11 vs_5_0 ps_5_0)
Major performance caveat: no
Maximum texture size: 16384
Point size range: 1 to 1024
Extensions:

EXT_color_buffer_float
EXT_disjoint_timer_query_webgl2
EXT_texture_filter_anisotropic
OES_texture_float_linear
WEBGL_compressed_texture_s3tc
WEBGL_compressed_texture_s3tc_srgb
WEBGL_debug_renderer_info
WEBGL_debug_shaders
WEBGL_lose_context
Audio information
System sample rate: 44100 Hz
Output channels: 2
Output interpretation: speakers
Supported decode formats:

Ogg Opus (audio/ogg; codecs=opus)
WebM Opus (audio/webm; codecs=opus)
Ogg Vorbis (audio/ogg; codecs=vorbis)
WebM Vorbis (audio/webm; codecs=vorbis)
MPEG-4 AAC (audio/mp4; codecs=mp4a.40.5)
MP3 (audio/mpeg)
FLAC (audio/flac)
PCM WAV (audio/wav; codecs=1)
Supported encode formats:

WebM Opus (audio/webm; codecs=opus)
Video information
Supported decode formats:

WebM VP9 (video/webm; codecs=vp9)
WebM VP8 (video/webm; codecs=vp8)
Ogg Theora (video/ogg; codecs=theora)
H.264 (video/mp4; codecs=avc1.42E01E)
Supported encode formats:

WebM VP9 (video/webm; codecs=vp9)
WebM VP8 (video/webm; codecs=vp8)
</details>

**Comment by el3um4s** [comment]
I compared a similar file saved with an old version of c3. In _layut 1.json_, there is more code than the sample file

`"nonworld-instances": [
		{
			"type": "JSON",
			"properties": {},
			"uid": 4,
			"instanceVariables": {}
		}
	],`

![nonworld-instances](https://user-images.githubusercontent.com/23579630/53818365-3d5a7d80-3f68-11e9-83a2-a17fc14462c1.PNG)

**Comment by shortercode** [comment]
How did you produce this example project? The issue is that the project has no instances of the JSON object, I'm not aware of any way of doing this actually inside construct.

**Comment by AshleyScirra** [comment]
I can reproduce with the provided project, but the built-in JSON example (that the project appears to be derived from) works correctly. So I guess you changed something in the example project that broke it.

**Comment by el3um4s** [comment]
@shortercode @AshleyScirra 

The file is the small version of a project I'm working on. The problem presented itself on a more complex project. I deleted all the elements and left the minimum to show the problem.

I did some tests and I think the bug is more general. I think the problem is related to global plugins.

I reproduced it by doing these steps

1. Insert a global plugin on a layout
2. Duplicate the layout
3. Delete the original layout

[bug.c3p.zip](https://github.com/Scirra/Construct-3-bugs/files/2936858/bug.c3p.zip)

This is a video that documents the various steps:
[See this video](https://www.youtube.com/watch?v=FpWCfl2D2fQ)

[![bug](https://img.youtube.com/vi/FpWCfl2D2fQ/0.jpg)](https://www.youtube.com/watch?v=FpWCfl2D2fQ)

**Comment by AshleyScirra** [comment]
Sounds like an issue with nonworld instances so taking assignment.

I still can't reproduce any problem. I opened the AJAX example, duplicated Layout 1 to Layout 2, deleted Layout 1, previewed Layout 2, and it still worked fine. I tried with a new project inserting a Dictionary object to Layout 1 and doing similar steps; the debugger showed there's still an instance of the Dictionary object on Layout 2. The attached project is not useful because it already has the JSON instance missing, and we need steps that cover it going missing in the first place.

The steps look like they should be trivial but the video is over 10 minutes long - please provide a shorter video or at least the relevant time range. Please also note that as described in the bug report guidelines, videos are not actually generally that useful - we need a minimal project and steps to reproduce that work.

**Comment by el3um4s** [comment]
[bug-short-video.c3p.zip](https://github.com/Scirra/Construct-3-bugs/files/2937136/bug-short-video.c3p.zip)

This is a shorter video, about 4 minutes.
[See this video](https://www.youtube.com/watch?v=LcmgIi9vjxI)

[![bug](https://img.youtube.com/vi/LcmgIi9vjxI/0.jpg)](https://www.youtube.com/watch?v=LcmgIi9vjxI)

The video is long because the problem occurs after a series of steps (for example after creating other layouts). I think there are no problems for very small projects. The steps seem trivial but occur after a few minutes of work.

By following the steps in the video you can repeatedly reproduce this problem.
Please watch the video, or, if you prefer, I will write here all the steps made in the video.

**Comment by el3um4s** [comment]
These are all the steps:

1. Create a new project
1. Add AJAX and JSON to _Layout 1_
1. Create a new JSON file
1. Add a New Layout with Event Sheet
1. Rename _Layout 2_ to _Menu_ and _Layout 1_ to _Test1_
1. Rename _Event Sheet 2_ to _Menu_ _Event Sheet 1_ to _Test1_
1. Add a _Text_ to Layout _Menu_
1. Add to event sheet _Menu_ the events: 
   * `+ System: On start of layout -> AJAX: Request New File.json (tag "")`
   * `+ AJAX: On "" completed -> JSON: Parse JSON string AJAX.LastData`
   * `+ AJAX: On "" completed -> Text: Set text to JSON.ToBeautifiedString`
1. Save the project as Project Folder
1. Close the project
1. Open the project
1. Duplicate _Test1_ (the new layout is _Test2_)
1. Duplicate _Test2_ (the new layout is _Test3_)
1. Duplicate _Test3_ (the new layout is _Test4_)
1. Add to event sheet _Test1_ the events: 
   * `+ System: On start of layout -> AJAX: Request New File.json (tag "")`
   * `+ AJAX: On "" completed -> JSON: Parse JSON string AJAX.LastData`
1. Save the project
1. Close the project
1. Open the project
1. Duplicate _Test3_ (the new layout is _Test5_)
1. Delete the layouts _Test1, Test2, Test3, Test4_
1. Save the project
1. Close the project
1. Open the project
1. Run the preview

**Comment by el3um4s** [comment]
I did some tests. It seems to me that the problem occurs only when I use the desktop version and I save the project as Project Folder

**Comment by AshleyScirra** [comment]
Can you find the smallest number of steps to reproduce the problem?

Folder saves work differently and it might be that some files are not being updated, so tagging as desktop.

**Comment by el3um4s** [comment]
I think this is the smallest number of steps to reproduce the problem.

I can not reproduce it without saving, closing and reopening the project several times.

I had this problem while I was reducing a project: that's why I deleted some layouts. Until I closed and reopened the project the day after I had no problems. 

I solved my problem by checking on github the differences between the various versions of my files

![cattura](https://user-images.githubusercontent.com/23579630/53911413-9ef91580-4056-11e9-9076-06b16c0939aa.PNG)

I am increasingly convinced that it is a strange bug and presents itself after a long series of operations. Or maybe it's me who does not have the skills and the knowledge to reduce the steps even further.
